### PR TITLE
Add timeout to the websocket connection

### PR DESCRIPTION
**Pull request by pcarranza**
When the Websocket connection locks for network reasons the whole thing just locks and provides no further output than

```
nlopes/slack2017/12/27 09:00:54 slack.go:91: Dialing to websocket on url wss://mpmulti-8ba0.lb.slack-msgs.com/websocket/KyN3bSd_30LXwzzJptxDwSH7ueeXerZuHa0qIX9IVmQpvepYC136bNxZfrEQiLCAreT6dtohZimnvM0q6H4GvkrSb3higukMj7iLJAwi9n0J_iEkhphlNo9Y4FyoPVhQeonNOVJC5KKQm-vgVypo_ifflS5IqSX-cLk5gp7GwRc=
```

With this PR the logs now continues with

```
2017/12/27 09:01:04 Received Slack Event &slack.ConnectionErrorEvent{Attempt:18, ErrorObj:(*websocket.DialError)(0x11664200)}
nlopes/slack2017/12/27 09:01:04 slack.go:91: Failed to dial to the websocket: websocket.Dial wss://mpmulti-8ba0.lb.slack-msgs.com/websocket/KyN3bSd_30LXwzzJptxDwSH7ueeXerZuHa0qIX9IVmQpvepYC136bNxZfrEQiLCAreT6dtohZimnvM0q6H4GvkrSb3higukMj7iLJAwi9n0J_iEkhphlNo9Y4FyoPVhQeonNOVJC5KKQm-vgVypo_ifflS5IqSX-cLk5gp7GwRc=: tls: DialWithDialer timed out
nlopes/slack2017/12/27 09:01:04 slack.go:91: reconnection 19 failed: websocket.Dial wss://mpmulti-8ba0.lb.slack-msgs.com/websocket/KyN3bSd_30LXwzzJptxDwSH7ueeXerZuHa0qIX9IVmQpvepYC136bNxZfrEQiLCAreT6dtohZimnvM0q6H4GvkrSb3higukMj7iLJAwi9n0J_iEkhphlNo9Y4FyoPVhQeonNOVJC5KKQm-vgVypo_ifflS5IqSX-cLk5gp7GwRc=: tls: DialWithDialer timed out
nlopes/slack2017/12/27 09:01:04 slack.go:97:  -> reconnecting in 5m0s
```

Which can be useful when running things on flaky connections, as connections can simply block forever for network weather reasons.

I also added a couple of debug log lines that helped me to troubleshoot the path to reach the websocket lockup from https://github.com/nlopes/slack/issues/226

### CODE REVIEW

**Comment by dvdplm** [comment]
lgtm but I'd like a second opinion as not very familiar with the managed WS code.

File: rtm.go
Comment (catherinetcai) [inline rtm.go]: Minor - can we move the timeout to a const?

Comment (pcarranza) [reply rtm.go]: Sure, done

File: rtm.go
Comment (catherinetcai) [inline rtm.go]: Why not just make this 10 * time.Second?

**Comment by james-lawrence** [comment]
LGTM once 10 * time.Second is fixed.

**Comment by james-lawrence** [comment]
actually there is one change we should do, but doesn't have to be part of this PR.

the net.Dialer should be configurable. the timeouts will vary depending on the network.

**Comment by pcarranza** [comment]
Ok, the change is done.

@james-lawrence I agree, all those timeouts should be configurable, I didn't want to go that far as I was debugging a networking problem and just wanted to have it solved. But it should totally be configurable.

Could you point me to what is the general way to do it? There is at least another timeout defined in the WebSocket proxy file (and a keepalive). I rather not pass a lot of arguments to get it done, but to use some form of configuration object, unless there's a better way.

**Comment by james-lawrence** [comment]
@pcarranza I would normally do it this way:

```golang

type RTMOption func(*RTM)

// RTMOptionWebsocketDialer set the dialer to be used when making websocket connections.
func RTMOptionWebsocketDialer(d net.Dialer) RTMOption {
	return func(rtm *RTM) {
		rtm.dialer = d
	}
}

// RTMOptionUseStart as of 11th July 2017 you should prefer setting this to false, see:
// https://api.slack.com/changelog/2017-04-start-using-rtm-connect-and-stop-using-rtm-start
func RTMOptionUseStart(b bool) RTMOption {
	return func(rtm *RTM) {
		rtm.useRTMStart = b
	}
}

// NewRTM returns a RTM, which provides a fully managed connection to
// Slack's websocket-based Real-Time Messaging protocol.
func (api *Client) NewRTM(option ...RTMOption) *RTM {
	result := &RTM{
		Client:           *api,
		IncomingEvents:   make(chan RTMEvent, 50),
		outgoingMessages: make(chan OutgoingMessage, 20),
		pings:            make(map[int]time.Time),
		isConnected:      false,
		wasIntentional:   true,
		killChannel:      make(chan bool),
		disconnected:     make(chan struct{}),
		forcePing:        make(chan bool),
		rawEvents:        make(chan json.RawMessage),
		idGen:            NewSafeID(1),
                dialer:           &net.Dialer{
			Timeout:   10 * time.Second,
			KeepAlive: 30 * time.Second,
	        	DualStack: true,
        	},
                useRTMStart: true,
	}
        
        for _, opt := range options {
                opt(r)
        }

        return r
}

// NewRTMWithOptions Deprecated just use NewRTM()
// returns a RTM, which provides a fully managed connection to
// Slack's websocket-based Real-Time Messaging protocol.
// This also allows to configure various options available for RTM API.
func (api *Client) NewRTMWithOptions(options *RTMOptions) *RTM {
	if options != nil {
		return api.NewRTM(RTMOptionUseStart(options.UseRTMStart))
	}
	return api.NewRTM()
}
```

```golang
func websocketProxyDial(d *net.Dialer, urlString, origin string) (ws *websocket.Conn, err error) {
	config, err := websocket.NewConfig(urlString, origin)
	if err != nil {
		return nil, err
	}
	config.Dialer = d

	if os.Getenv("HTTP_PROXY") == "" {
		return websocket.DialConfig(config)
	}

	// ...

	return websocket.NewClient(config, client)
}
```

**Comment by james-lawrence** [comment]
added a way to do it without breaking backwards compatability with current NewRTM methods.

File: websocket_managed_conn.go
Comment (james-lawrence) [inline websocket_managed_conn.go]: this comment needs to remain.

File: rtm.go
Comment (james-lawrence) [inline rtm.go]: defaultWebsocketDialTimeout, no need to clutter up the exposed API with immutable internal constants.

**Review by james-lawrence — CHANGES_REQUESTED** [review]
small changes required.

if you want to do the changes to NewRTM() I mentioned prior that'd be wonderful.

**Review by james-lawrence — APPROVED** [review]

**Pull request merged**
